### ISSUE TITLE
Dashes in IAM user name broken with S3 backend

### ISSUE THREAD

**Issue by jeffb4** [hashicorp/terraform-provider-aws#9550]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

```
Terraform v0.12.5
+ provider.aws v2.21.1
```

### Affected Resource(s)

* aws_iam_user

### Terraform Configuration Files

```hcl
terraform {
  required_version = ">= 0.12.0"

  required_providers {
    aws        = "~> 2.21"
  }
  backend "s3" {}
}

provider "aws" {
  region = "us-east-1"
}

resource "aws_iam_user" "dash_test" {
  name          = "dash-test"
  force_destroy = true
  path          = "/"
}
```

### Debug Output

(redacted) debug output with dashes in username https://gist.github.com/jeffb4/f0fccab91495597b622b2097163c881b

(redacted) debug output without dashes in username https://gist.github.com/jeffb4/ab932903779a2758e72aaea151f98f72

### Expected Behavior

User should have been created with a dash in name. This behavior works perfectly fine when the backend is not S3 (and yes I will be opening an issue on terraform/terraform for it)

### Actual Behavior

![image](https://user-images.githubusercontent.com/3298329/62145924-90372780-b2b1-11e9-855e-5b839967c0d3.png)

user is not in "None" for Groups

![image](https://user-images.githubusercontent.com/3298329/62145984-a9d86f00-b2b1-11e9-8c11-a5c1bbce0a07.png)

user shows "Entity not found" in IAM console

### Steps to Reproduce

1. Create tf file as above
2. Configure state backend with `terraform init -backend-config`
3. `terraform apply`

### Important Factoids

Using AWS SSO for an assumed role for runs

### References

https://github.com/hashicorp/terraform/issues/22257

**Comment by jeffb4** [comment]
I've also seen this problem with underscores `_` as well, now. No special characters still seems to work fine.

**Comment by nywilken** [comment]
Hi @jeffb4 can you elaborate a little more on what exactly is the problem you are running into. You mention S3 backend being broken, the creation of the `aws_iam_user` in the provided configuration is independent of any S3 backend related operations so the user should be created without issue.

When running `terraform init -backend-config` do you run into any errors? Can you access the state file being referenced within the S3 backend bucket?

I have validated using the provided configuration that user creation works as expected for both dashes and underscores.

**Comment by jeffb4** [comment]
The problem is that when I was trying to create an IAM user with a dash `-` in the middle of its name using the S3 backend, it was creating a broken unusable user (I'm not sure how).

I understand that creation should be independent of backend used, however my user creation was successful with local state file and broken with S3.

I ran into no errors during init. I can access the state file being referenced.

Do the debug logs I uploaded indicate anything?
